apps/designer/app/routes/$.tsx

```diff
@@ -0,0 +1,128 @@
+import { type LoaderFunction, redirect, MetaFunction } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { InstanceRoot, Root } from "@webstudio-is/react-sdk";
+import { loadCanvasData, type CanvasData } from "~/shared/db";
+import config from "~/config";
+import { db } from "@webstudio-is/project/index.server";
+import env, { Env } from "~/env.server";
+import { sentryException } from "~/shared/sentry";
+import { Canvas } from "~/canvas";
+
+type Mode = "edit" | "preview" | "published";
+
+type Data =
+  | (CanvasData & { env: Env; mode: Mode })
+  | { errors: string; env: Env };
+
+type ProjectIdentifier = { type: "id" | "domain"; value: string };
+
+export const isCanvasRequest = (
+  request: Request
+): false | { projectId: ProjectIdentifier; mode: Mode; pathname: string } => {
+  const url = new URL(request.url);
+
+  const pathname = url.pathname;
+
+  if (Object.values(config).some((path) => pathname.startsWith(path))) {
+    return false;
+  }
+
+  let projectId: ProjectIdentifier | undefined = undefined;
+
+  // @todo all this subdomain logic is very hacky
+  const host =
+    request.headers.get("x-forwarded-host") ||
+    request.headers.get("host") ||
+    "";
+  const [userDomain, wstdDomain] = host.split(".");
+  if (wstdDomain === "wstd" || wstdDomain?.includes("localhost")) {
+    projectId = { type: "domain", value: userDomain };
+  }
+
+  if (projectId === undefined) {
+    const projectIdParam = url.searchParams.get("projectId");
+    if (projectIdParam !== null) {
+      projectId = { type: "id", value: projectIdParam };
+    }
+  }
+
+  if (projectId === undefined) {
+    return false;
+  }
+
+  const modeParam = url.searchParams.get("mode");
+
+  if (modeParam === "edit") {
+    return { projectId, mode: "edit", pathname };
+  }
+
+  if (modeParam === "preview") {
+    return { projectId, mode: "preview", pathname };
+  }
+
+  return { projectId, mode: "published", pathname };
+};
+
+export const meta: MetaFunction = ({ data }: { data: Data }) => {
+  if ("errors" in data) {
+    return { title: "Error" };
+  }
+  const { page } = data;
+  return { title: page.title, ...page.meta };
+};
+
+export const loader: LoaderFunction = async ({
+  request,
+}): Promise<Data | Response> => {
+  try {
+    const canvasRequest = isCanvasRequest(request);
+
+    if (canvasRequest === false) {
+      return redirect(config.dashboardPath);
+    }
```

Comment:
if its not a request for canvas, shouldn't we yield to the router to handle it? What happens if its some other route e.g. /designer/something ?

Reply:
This is for the case when you open `localhost:3000`, it just redirects you to dashboard. This logic is copied from `index.tsx`

Reply:
will it allow us having a route like /designer/something ? with the normal route handling

Reply:
Sure. Remix fallbacks to `$.tsx` only if it didn't find any other matching route.